src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +151,150 @@ internal class DiagnosticComputer
+            //  - Note that it is critical to do this step prior to Step 3 below to ensure that
+            //    any canceled normal priority tasks in Step 3 do not resume execution prior to
+            //    completion of this high priority computeTask. 
+            lock (s_gate)
+            {
+                Debug.Assert(!s_highPriorityComputeTasks.Contains(computeTask));
+                s_highPriorityComputeTasks = s_highPriorityComputeTasks.Add(computeTask);
+            }
+
+            try
+            {
+                // Step 3:
+                //  - Force cancellation of all the executing normal priority tasks
+                //    to minimize resource and CPU contention between normal priority tasks
+                //    and the high priority computeTask in Step 4 below.
+                CancelNormalPriorityTasks();
+
+                // Step 4:
+                //  - Execute the core 'computeTask' for diagnostic computation.
+                return await computeTask.ConfigureAwait(false);
+            }
+            finally
+            {
+                // Step 5:
+                //  - Remove the 'computeTask' from the list of current executing high priority tasks.
+                lock (s_gate)
+                {
+                    Debug.Assert(s_highPriorityComputeTasks.Contains(computeTask));
+                    s_highPriorityComputeTasks = s_highPriorityComputeTasks.Remove(computeTask);
+                }
+            }
+
+            static void CancelNormalPriorityTasks()
+            {
+                // Fire cancellation on the current token source and replace it with a new token source for future normal priority requests.
+                CancellationTokenSource currentCts;
+                lock (s_gate)
+                {
+                    currentCts = s_cancellationSourceForNormalPriorityComputeTasks;
+                    s_cancellationSourceForNormalPriorityComputeTasks = new();
+                }
+
+                currentCts.Cancel();
+                currentCts.Dispose();
+            }
+        }
+
+        private async Task<SerializableDiagnosticAnalysisResults> GetNormalPriorityDiagnosticsAsync(
+            IEnumerable<string> analyzerIds,
+            bool reportSuppressedDiagnostics,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Step 1:
+                //  - Normal priority task must wait for all the executing high priority tasks to complete
+                //    before beginning execution.
+                await WaitForHighPriorityTasksAsync(cancellationToken).ConfigureAwait(false);
+
+                // Step 2:
+                //  - Create the core 'computeTask' for computing diagnostics.
+                var computeTask = CreateComputeTask(cancellationToken);
```

Comment:
nit: this task isn't stored anwhere.  just await the normal work within the try/catch.